Serialization/ArgsList.Serialization.cs

```diff
@@ -0,0 +1,182 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using UnityEngine;
+
+namespace DUCK.Serialization
+{
+	public partial class ArgsList
+	{
+		private const string COMPONENT_PREFIX = "c:";
+		
+		[SerializeField]
+		private string[] typeOrder;
+
+		[SerializeField]
+		private string[] stringValues;
+
+		[SerializeField]
+		private int[] intValues;
+
+		[SerializeField]
+		private float[] floatValues;
+
+		[SerializeField]
+		private bool[] boolValues;
+
+		[SerializeField]
+		private GameObject[] gameObjectArgs;
+
+		[SerializeField]
+		private Vector2[] vector2Args;
+
+		[SerializeField]
+		private Vector3[] vector3Args;
+
+		[SerializeField]
+		private Vector4[] vector4Args;
+
+		[SerializeField]
+		private Color[] colorArgs;
+
+		public void OnBeforeSerialize()
+		{
+			if (argTypes == null) return;
+
+			var argLists = new Dictionary<Type, List<object>>();
+			var typeOrderList = new List<string>();
+
+			Func<Type, List<object>> lazyGetList = (t) =>
+			{
+				if (t.IsSubclassOf(typeof(Component)))
+				{
+					t = typeof(GameObject);
+				}
+				if (argLists.ContainsKey(t))
+				{
+					return argLists[t];
+				}
+				return argLists[t] = new List<object>();
+			};
+
+			for (int i = 0; i < argTypes.Count; i++)
+			{
+				var argType = argTypes[i];
+				var list = lazyGetList(argType);
+				if (argType.IsSubclassOf(typeof(Component)))
+				{
+					var arg = args[i];
+					list.Add(arg);
+					typeOrderList.Add(COMPONENT_PREFIX + argType.Name);
+				}
+				else
+				{
+					list.Add(args[i]);
+					typeOrderList.Add(argType.Name);
+				}
+			}
+
+			typeOrder = typeOrderList.ToArray();
+
+			foreach (var supportedType in supportedTypes)
+			{
+				supportedType.Value.SetList(this, lazyGetList(supportedType.Value.Type));
+			}
+		}
+
+		public void OnAfterDeserialize()
+		{
+			var serializedLists = new Dictionary<string, List<object>>();
+			args = new List<object>();
+			var localArgTypes = new List<Type>();
+
+			// Convert all serialized arrays to lists.
+			foreach (var supportedType in supportedTypes.Values)
+			{
+				var list = supportedType.GetList(this);
+				serializedLists.Add(supportedType.Type.Name, list);
+			}
+
+			for (var i = 0; i < typeOrder.Length; i++)
+			{
+				var typeName = typeOrder[i];
+				List<object> list;
+				Type argType;
+
+				if (typeName.StartsWith(COMPONENT_PREFIX))
+				{
+					// strip off prefix
+					typeName = typeName.Replace(COMPONENT_PREFIX, "");
+					if (!componentTypes.ContainsKey(typeName))
+					{
+						// TODO: How to handle this error (found a component type not in the assemblies on deserialize)
```

Comment:
For now I've just put to throw, but dunno if there's a better option

We support all component types for unity assembly, which contains all the built in components (Renderer, collider camera etc...) and the project's main assembly*. If you try save a component form another assembly, like some third party DLL or something it will throw.


*More specifically, the assembly that contains the ArgsList type. Given that duck is distributed in a submodule it's gonna be the project's main assembly, unless you do something weird or we ever decide to distribute duck as it's own assembly. (if we did that this would need rethinking)